src/spotify_to_tidal/sync.py

```diff
@@ -166,12 +172,19 @@ def get_tracks_from_spotify_playlist(spotify_session: spotipy.Spotify, spotify_p
             return output
 
 class TidalPlaylistCache:
```

Comment:
After making this a singleton, it's not really a `TidalPlaylistCache` anymore and more of a `TidalTrackCache`.

The original purpose of the cache was just to avoid searching for ALL tracks in the current spotify playlist, and rather just search for the new tracks in the current spotify playlist. If I understand correctly, you now want to accumulate all the Tidal tracks over the course of the sync operation, which can save time when you have a lot of track duplication across different playlist? If so then it's probably a good idea, but of course it would come at the cost of increased memory usage. Would you be able to share a rough number for how much memory each Tidal track uses?

Reply:
Yeah, the purpose is to keep the cache persistent through the run. if a user is syncing multiple playlist and share the same songs, it will retrieve it from the cache reducing a network call. I can change the naming if this feature is to stay

I ran `sys.getsizeof()` on `TidalPlaylistCache._data` with an 100 song playlist, it's showing usage of 8408 bytes

Reply:
That would just be 84 bytes per track which doesn't sound right, it's probably just the size of the data structure without considering the size of the underlying json data.

Reply:
I found this function online:

```python
import sys

def get_size(obj, seen=None):
    """Recursively finds size of objects"""

    size = sys.getsizeof(obj)
    if seen is None:
        seen = set()

    obj_id = id(obj)
    if obj_id in seen:
        return 0

    # Important mark as seen *before* entering recursion to gracefully handle
    # self-referential objects
    seen.add(obj_id)

    if isinstance(obj, dict):
        size += sum([get_size(v, seen) for v in obj.values()])
        size += sum([get_size(k, seen) for k in obj.keys()])
    elif hasattr(obj, '__dict__'):
        size += get_size(obj.__dict__, seen)
    elif hasattr(obj, '__iter__') and not isinstance(obj, (str, bytes, bytearray)):
        size += sum([get_size(i, seen) for i in obj])

    return size
```

That gives about 71kB per track

Reply:
I have about 10,000 tracks in my collection, so that would accumulate to ~700MB. I guess it's ok but if someone had say 100,000 tracks it might start to be an issue.

Reply:
Yeah. I'm trying to profile the code with scalene to verify, but I ran into some WSL issues. Currently trying to fix those issues

Reply:
The bigger problem is that the search algorithm in the cache is O(N) on the size of the tracks in the cache:

```
    def _search(self, spotify_track: t_spotify.SpotifyTrack) -> tidalapi.Track | None:
        ''' check if the given spotify track was already in the tidal playlist.'''
        for tidal_track in self._data:
            if match(tidal_track, spotify_track):
                return tidal_track
        return None
```

So this would slow down considerably for larger collections, and probably not worth it, which I think is why I didn't do this in the first place

Reply:
What could be a useful addition is to introduce a second cache which just contains a lightweight dictionary that maps the spotify track id to a tidal track id (i.e. `str` -> `str` dictionary). You could accumulate this for every track in the collection and probably even persist to disk. You'd probably still want to keep the existing `PlaylistCache` which contains the full Tidal objects for each track, a bit hard to say without actually trying it and measuring the performance.

Reply:
Yeah, makes sense. With the `cachetools` usage from my fork, I _want to_ say it achieves the same goal as the `TidalPlaylistCache`. I'll test different implementations and see which is best

Reply:
Removing the cache class all together and relying on cachetools resulted in 22% faster runtime.

Instead of caching the tracks prior to running. Maybe the lightweight cache mapping spotify ids -> tidal ids upon track match would be better?

Reply:
Removing the cache class all together is not a good idea imo. Consider the following case:

1. You haven't run the script for a few days+ (TTL has expired)
2. You have fairly large playlists (100+ tracks) each
3. The spotify playlists have not changed much since last time you ran the script
4. You have not manually modified the Tidal playlists created by the script last time

The existing playlist cache class addresses this case by loading the tracks from each tidal playlist at the start of the sync operation, and skipping tracks which are already known (from the last sync operation, which was potentially months ago). If you only have a TTL cache this case would be significantly slowed down

Reply:
Ah okay. I changed how `TidalTrackCache` stores data. It now maps IDs with a map instead of storing objects to reduce memory. When populating the cache and a match is found, I remove the tidal track from possible matches. This will reduce memory usage and search time.

Reply:
LGTM!